---
title: "variant calling markdown"
output: html_document
date: "2024-03-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Step 1: Demultiplexing

Demultiplexing of pooled samples using sabre. Barcode files should be prepared as described in https://github.com/najoshi/sabre. 32 samples per file, 3 paired end files for each plate = 6 files per plate.

```{bash}

#!/bin/bash
##SBATCH -p medium
##SBATCH -t 10:00:00
##SBATCH -N 4
##SBATCH -c 8

sabre pe -c -f /usr/users/aiyesa/geno_fastq_data/fastq/2023_data/M005083-Plate8_S136_L003_R1_001.fastq.gz \
-r /path/to/raw_data/M005083-Plate8_S136_L003_R2_001.fastq.gz \
-b /path/to/raw_data/first_bcode.txt \
-u /path/to/raw_data/demultiplexed_samples/unknown_barcode8f_1_32.fastq \
-w /path/to/raw_datademultiplexed_samples/unknown_barcode8r_1_32.fastq

```

## Step 2: Read quality assessment

```{bash}

#!/bin/bash

module load FASTQC/0.11.4

fastqc -t 8 /path/to/demultiplexed_samples/zipfiles/*.gz -o fastqc_result

# multiqc *.gz ## using multiqc to aggregate output from fastqc output

```

## Step 3: Trimming adapters

```{bash}

#!/bin/bash
#SBATCH -J fastq_trimming
#SBATCH -o gbs-%j.out
#SBATCH -N 4
#SBATCH -t 2-00:00
#SBATCH -n 16


INDS=($(for i in /path/to/demultiplexed_samples/zipfiles/*_R1.fastq.gz; do echo $(basename "$i" _R1.fastq.gz); done))

### run through samples
for IND in ${INDS[@]};
do
	# declare variables
	FORWARD=/path/to/demultiplexed_samples/zipfiles/${IND}_R1.fastq.gz
	REVERSE=/path/to/demultiplexed_samples/zipfiles/${IND}_R2.fastq.gz
	OUTFOR=/usr/users/aiyesa/geno_fastq_data/fastq/2023_data/trimmed_data/${IND}_trimmed_R1.fastq.gz
	OUTREV=/usr/users/aiyesa/geno_fastq_data/fastq/2023_data/trimmed_data/${IND}_trimmed_R2.fastq.gz

	# then align and sort
	echo "trimming $IND with cutadapt"
	cutadapt -a AGATCGGAAGAGC -A AGATCGGAAGAGC --core 0 $FORWARD $REVERSE -o $OUTFOR -p $OUTREV ;
done

```

## Step 4: Mapping to reference genom

```{bash}

#!/bin/bash
#SBATCH --job-name=alignment
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=8
#SBATCH --mem=32G
#SBATCH --time=48:00:00
#SBATCH --output=alignment.%A.%a.out
#SBATCH --error=alignment.%A.%a.err
#SBATCH --array=0-34%32   # 35 batches of 32 samples each (0-34 with step 32)

# Load necessary modules
module load bwa
module load samtools

# Set your input and output directories
INPUT_DIR=/path/to/trimmed_data
OUTPUT_DIR=/path/to/sorted_bamfiles
REFERENCE_GENOME=/path/to/reference/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna_sm.toplevel.fa

# List of sample names
SAMPLES=($(for i in /path/to/working_dir/trimmed_data/*_R1.fastq.gz; do echo $(basename "$i" _R1.fastq.gz); done))

# Calculate the start and end index for the current batch
start_index=$((SLURM_ARRAY_TASK_ID * 32))
end_index=$((start_index + 31))

# Process each sample in the current batch
for ((i = start_index; i <= end_index && i < ${#SAMPLES[@]}; i++)); do
    CURRENT_SAMPLE=${SAMPLES[i]}

    # BWA alignment and pipe to samtools for sorting
    bwa mem -t 8 $REFERENCE_GENOME $INPUT_DIR/${CURRENT_SAMPLE}_R1.fastq.gz $INPUT_DIR/${CURRENT_SAMPLE}_R2.fastq.gz | \
    samtools view -@ 8 -S -b | \
    samtools sort -@ 8 -o $OUTPUT_DIR/${CURRENT_SAMPLE}_aligned_sorted.bam

    # Index BAM file
    samtools index $OUTPUT_DIR/${CURRENT_SAMPLE}_aligned_sorted.bam

    # Clean up intermediate files if desired
    rm $OUTPUT_DIR/${CURRENT_SAMPLE}_aligned.bam

    echo "Alignment for $CURRENT_SAMPLE completed successfully."
done

echo "Batch completed successfully."

```

## Step 5: Variant calling

```{bash}

#!/bin/bash
#SBATCH --job-name=variant-calling
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --cpus-per-task=8
#SBATCH --mem=32G
#SBATCH --time=48:00:00
#SBATCH --output=variant-calling.%A.%a.out
#SBATCH --error=variant-calling.%A.%a.err
##SBATCH --array=0-34%32   # 35 batches of 32 samples each (0-34 with step 32)

module load samtools
module load bcftools

REF=/path/to/working_dir/reference/ref5.0/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna_sm.toplevel.fa

bcftools mpileup -a AD,DP,SP -Ou -f $REF \
/path/to/working_dir/sorted_bamfiles/*_sorted.bam | bcftools call -f GQ,GP \
-mO z -o zea_individual_plant_geno.vcf.gz

```

## Step 6: FIltering 

### Filtering parameters include missingness of 20%, minor allele frequency of 0.01, QUAL score of 30, min depth of 5, and, maximum depth of 500.

```{bash}

#!/bin/bash
#SBATCH --job-name=vcf_filtering
#SBATCH --output=vcf_filtering_%A_%a.out
#SBATCH --error=vcf_filtering_%A_%a.err
#SBATCH --array=1-10  # Number of VCF files / chromosomes
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=4G
#SBATCH --time=48:00:00

# Set parameters
MAF="0.01"
MISS="0.2"
QUAL="30"
MIN_DEPTH="2"
MAX_DEPTH="500"

# Load necessary modules (if required)
module load vcftools

# Define input and output paths
VCF_IN="/path/to/working_dir/sorted_bamfiles/filtered_data/zea_geno_chr${SLURM_ARRAY_TASK_ID}.vcf.gz"
VCF_OUT="/path/to/working_dir/filtered_data/filter_2min_500max/zea_geno_chr${SLURM_ARRAY_TASK_ID}.filtered.vcf.gz"

# Perform filtering with vcftools
vcftools --gzvcf "$VCF_IN" \
    --remove-indels --maf "$MAF" --max-missing "$MISS" --minQ "$QUAL" \
    --min-meanDP "$MIN_DEPTH" --max-meanDP "$MAX_DEPTH" \
    --minDP "$MIN_DEPTH" --maxDP "$MAX_DEPTH" --recode --stdout | gzip -c > "$VCF_OUT"

```

